### PR TITLE
[auth] feat: axum with routes

### PR DESCRIPTION
**Pull request by chesedo**

### CODE REVIEW

File: gateway/src/api/latest.rs [hunk 1]
```diff
@@ -436,36 +436,16 @@ impl ApiBuilder {
...
-                        debug!(
-                            latency = format_args!("{} ns", latency.as_nanos()),
-                            "finished processing request"
-                        );
-                    },
-                ),
+            TraceLayer::new(|request| {
+                request_span!(
+                    request,
+                    account_name = field::Empty,
```
Comment (chesedo) [inline gateway/src/api/latest.rs:442]:
```suggestion
                    account.name = field::Empty,
```

File: common/src/backends/metrics.rs [hunk 2]
```diff
@@ -33,3 +41,279 @@ where
...
+            path.as_str()
+        } else {
+            ""
+        };
+
+        tracing::debug_span!(
+            "request",
+            http.uri = %$request.uri(),
+            http.method = %$request.method(),
+            http.status_code = tracing::field::Empty,
```
Comment (oddgrd) [inline common/src/backends/metrics.rs:186]: Doesn't the comment for this macro indicate we will record the http status?

File: common/src/backends/metrics.rs [hunk 3]
```diff
@@ -33,3 +41,279 @@ where
...
+
+impl TraceLayer<MakeSpanSimple> {
+    /// Switch to the span maker which does not add propagation details from the request headers
+    pub fn without_propagation(self) -> Self {
+        self
+    }
+}
+
+impl TraceLayer<MakeSpanPropagation> {
+    /// Switch to the span maker which adds propagation details from the request headers
```
Comment (oddgrd) [inline common/src/backends/metrics.rs:98]: Tracing noob question: it means propagation from request headers to response headers, right?

**Review by oddgrd — APPROVED** [review]
LGTM! I need to work on it for a bit to get a complete understanding, but I think this is good to go. I added a few questions I had.

The failing test works for me locally, maybe it needs a longer sleep in CI? :shrug:

File: common/src/backends/metrics.rs (hunk 2 above)
Comment (chesedo) [reply common/src/backends/metrics.rs:186]: Yip, however we'll only know the status once the request is complete. This happens on the [OnResponseStatusCode](https://github.com/shuttle-hq/shuttle/pull/632/files#diff-f4cf651c8b021e33d0f88c58c9b6648ff6db38f596dbb9130eaa5163a8fd4880R159)

File: common/src/backends/metrics.rs (hunk 3 above)
Comment (chesedo) [reply common/src/backends/metrics.rs:98]: Only from the request headers. If we have an HTTP client that calls another service (like gateway calling the deployers), then we have to manually set these propagation details on the headers (currently)

Comment (chesedo) [reply common/src/backends/metrics.rs:98]: Here's an example of the headers being set
https://github.com/shuttle-hq/shuttle/blob/b3e11b5ec0eb129dfbe06f90b99c785ea5334126/gateway/src/proxy.rs#L157-L159

Comment (chesedo) [reply common/src/backends/metrics.rs:98]: So `MakeSpanPropagation` will set it from the request headers to the tracing context (for the parent span) to be more precise

Comment (oddgrd) [reply common/src/backends/metrics.rs:98]: Thanks!

**Pull request merged**